### ISSUE TITLE
[AMD][kernel][Tier 6] Port sol-execbench L2/072_region_aware_self_attention_with_edit_bias_backward to FlyDSL

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#262]
# FlyDSL implementation: port sol-execbench's `072_region_aware_self_attention_with_edit_bias_backward` problem to FlyDSL

## Goal

Port the `072_region_aware_self_attention_with_edit_bias_backward` problem from `nvidia/SOL-ExecBench` (subset `L2`) to FlyDSL as part of the sol-execbench → FlyDSL coverage sweep. This is one of 235 problems across 4 subsets (L1, L2, Quant, FlashInfer-Bench) being tracked for FlyDSL kernel coverage.

**Source:** `stepfun-ai/Step1X-Edit-v1p2` (per `definition.json` `hf_id`)

**Description:** Backward pass for region-aware self-attention with edit bias. Computes gradients through QKV projection, scaled dot-product attention with position-based and edit-region biases, softmax, dropout, and output projection.

## Hardware / workload

```text
Hardware:  1 × MI300X (gfx942) + 1 × MI355X (gfx950)
Models:    (none — kernel benchmark, not LLM workload)
Subset:    L2
Problem:   072_region_aware_self_attention_with_edit_bias_backward
Source:    stepfun-ai/Step1X-Edit-v1p2
```

**Const axes:**
  - `hidden_size` = 1024
  - `num_attention_heads` = 16
  - `head_dim` = 64
  - `max_position_embeddings` = 1024

**Inputs:**
  - `grad_output` [batch_size, seq_len, hidden_size] float32 — Gradient w.r.t. attention output
  - `hidden_states` [batch_size, seq_len, hidden_size] float32 — Input hidden states from forward pass
  - `edit_region_mask` [batch_size, seq_len] float32 — Binary mask indicating edit regions
  - `qkv_weight` [qkv_size, hidden_size] float32 — QKV projection weight matrix
  - `qkv_bias` [qkv_size] float32 — QKV projection bias
  - `out_weight` [hidden_size, hidden_size] float32 — Output projection weight matrix
  - `out_bias` [hidden_size] float32 — Output projection bias
  - `edit_region_bias` [num_attention_heads, max_position_embeddings, max_position_embeddings] float32 — Position-based edit region bias
  - `within_edit_bias` [num_attention_heads, 1, 1] float32 — Bias for within-edit attention
  - `cross_edit_bias` [num_attention_heads, 1, 1] float32 — Bias for cross-edit attention
  - `attention_mask` [batch_size, seq_len] float32 — Attention mask (1 for valid, 0 for masked)
  - `query` [batch_size, num_attention_heads, seq_len, head_dim] float32 — Query tensor from forward pass
  - `key` [batch_size, num_attention_heads, seq_len, head_dim] float32 — Key tensor from forward pass
  - `value` [batch_size, num_attention_heads, seq_len, head_dim] float32 — Value tensor from forward pass
  - `attention_scores` [batch_size, num_attention_heads, seq_len, seq_len] float32 — Attention scores from forward pass
  - `attention_probs` [batch_size, num_attention_heads, seq_len, seq_len] float32 — Attention probabilities from forward pass
  - `attention_probs_dropped` [batch_size, num_attention_heads, seq_len, seq_len] float32 — Attention probabilities after dropout
  - `attention_output` [batch_size, seq_len, hidden_size] float32 — Attention output before final projection
  - `dropout_mask` [batch_size, num_attention_heads, seq_len, seq_len] float32 — Dropout mask from forward pass
  - `scale` [] float32 — Attention score scaling factor
  - `dropout_p` [] float32 — Dropout probability

**Outputs:**
  - `grad_hidden_states` [batch_size, seq_len, hidden_size] float32 — Gradient w.r.t. hidden states
  - `grad_qkv_weight` [qkv_size, hidden_size] float32 — Gradient w.r.t. QKV weight
  - `grad_qkv_bias` [qkv_size] float32 — Gradient w.r.t. QKV bias
  - `grad_out_weight` [hidden_size, hidden_size] float32 — Gradient w.r.t. output weight
  - `grad_out_bias` [hidden_size] float32 — Gradient w.r.t. output bias
  - `grad_edit_region_bias` [num_attention_heads, max_position_embeddings, max_position_embeddings] float32 — Gradient w.r.t. edit region bias
  - `grad_within_edit_bias` [num_attention_heads, 1, 1] float32 — Gradient w.r.t. within-edit bias
  - `grad_cross_edit_bias` [num_attention_heads, 1, 1] float32 — Gradient w.r.t. cross-edit bias

**Workload samples (first 3):**
  - {"uuid": "c3e6b5a0-3af9-5db6-b58e-e8419c2ff160", "axes": {"batch_size": 64, "seq_len": 128}, "inputs": {"grad_output": {"type": "custom"}, "hidden_states": {"type": "custom"}, "edit_region_mask": {"ty
  - {"uuid": "42bf7827-0cf3-56bb-87cf-26aea2de0058", "axes": {"batch_size": 1, "seq_len": 373}, "inputs": {"grad_output": {"type": "custom"}, "hidden_states": {"type": "custom"}, "edit_region_mask": {"typ
  - {"uuid": "4bc6fe79-aeed-54a6-b0bd-30a500b125f1", "axes": {"batch_size": 32, "seq_len": 293}, "inputs": {"grad_output": {"type": "custom"}, "hidden_states": {"type": "custom"}, "edit_region_mask": {"ty

```text
Baseline:  sol-execbench PyTorch reference (reference.py) on AMD with torch.compile off
Target:    max_abs_diff < 1e-2 vs PyTorch reference on all tested shapes
           kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
```

## Scope

1. `kernels/sol_execbench/l2_072_region_aware_self_attention_with_edit_bias_backward.py` — FlyDSL kernel implementing the op described in `definition.json`, matching the PyTorch reference's numerics
2. `tests/kernels/test_sol_execbench_l2_072_region_aware_self_attention_with_edit_bias_backward.py` — parity harness vs PyTorch reference, testing multiple workload shapes
3. `tests/kernels/bench_sol_execbench_l2_072_region_aware_self_attention_with_edit_bias_backward.py` — reports `kernel_ms` per shape + comparison vs PyTorch eager baseline
4. `assessment.md` (≤300 words) — ranking which shapes FlyDSL wins/ties/loses vs PyTorch eager

## Acceptance

- Correctness: max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes
- Performance: kernel_ms < 0.9 × PyTorch eager baseline on ≥80% of workload shapes
- Coverage: works on both gfx942 (MI300X) and gfx950 (MI355X)

## Out of scope

- Backward pass (unless the problem is itself a backward op)
- Weight gradients
- Optimizer step
- Cross-arch comparison vs NVIDIA H100 (PyTorch eager on AMD is the baseline)

## Reproduction

Clone `amdpilot-org/SOL-ExecBench`, download the dataset (`scripts/download_data.sh`), find the problem at `data/benchmark/L2/072_region_aware_self_attention_with_edit_bias_backward/`, run the sol-execbench driver against a FlyDSL solution. Requires the unified aiter+triton+flydsl Docker image at `/home/xiasun/unified-rocm.Dockerfile` (not yet built).

## Definition of done

- [ ] Kernel implemented at `kernels/sol_execbench/l2_072_region_aware_self_attention_with_edit_bias_backward.py`
- [ ] Tests pass on MI300X (gfx942)
- [ ] Tests pass on MI355X (gfx950)
- [ ] Bench shows ≥80% win rate vs PyTorch eager
- [ ] `assessment.md` written

## ## Leaderboard baseline to beat

These are the current best public results on NVIDIA B200 (from the
[sol-execbench leaderboard](https://research.nvidia.com/benchmarks/sol-execbench/leaderboard))
for problem `072_region_aware_self_attention_with_edit_bias_backward` (kernel_id=166). The agent's
FlyDSL kernel on MI300X/MI355X should aim to beat the **Scoring Baseline**
(optimized PyTorch) and approach the **top human/AI submissions**, normalized
by cross-arch roofline.

```text
GPU:           NVIDIA B200 (leaderboard) vs AMD MI300X/MI355X (this task)
Latency unit:  ms

SOL Bound (theoretical roofline):     0.038346 ms   (SOL-Score=1.0000, speedup=30.67×)
Scoring Baseline (optimized PyTorch): N/A ms   (SOL-Score=N/A, speedup=N/A×)
Reference Implementation (plain PyTorch): 2.621997 ms   (SOL-Score=0.2985, speedup=0.45×)

Top public submissions:
  - **Rank 1** (`Amir M. Mir | SF Tensor`): 0.335250 ms, SOL-Score=0.7829, speedup=3.39× vs reference
  - **Rank 2** (`Recursive`): 0.393252 ms, SOL-Score=0.7518
  - **Rank 3** (`doubleAI`): 0.413231 ms, SOL-Score=0.7412
```

**Cross-arch note:** B200 timings are not directly comparable to MI300X/MI355X
(different GPU arch, different roofline). The agent should:
1. Measure the PyTorch reference on MI300X/MI355X (same shapes, same Docker image)
2. Compute the FlyDSL kernel's speedup vs the on-AMD PyTorch reference
3. Compare that speedup ratio to the leaderboard's `avg_speedup` ratios above
4. Target: FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × (Rank-1 speedup vs PyTorch-reference-on-B200)

## Definition of done

- [ ] FlyDSL kernel passes correctness (max_abs_diff < 1e-2 vs PyTorch reference on all tested workload shapes)
- [ ] FlyDSL kernel_ms measured on MI300X (gfx942) AND MI355X (gfx950) across >=5 workload shapes
- [ ] FlyDSL speedup vs PyTorch-reference-on-AMD >= 0.8 × Rank-1 speedup vs PyTorch-reference-on-B200 (=3.39× if top1 exists)
- [ ] `assessment.md` (<=300 words) ranks FlyDSL vs PyTorch eager on each shape, identifies win/tie/lose regimes
- [ ] Patch modifies actual kernel source (not just bench scripts) — bench-only patches are REJECTED

## Non-goals / rejection criteria

- Bench-only patches (no FlyDSL kernel source changes) are rejected
- Patches that set expected_value = measured_value (tautological pass) are rejected
- The metric contract must compare FlyDSL vs PyTorch reference on AMD, not an absolute ms
- "Completed" without a measured speedup ratio vs the on-AMD PyTorch reference is rejected

**Comment by sunxxuns** [comment]
## AMDPilot acceptance target

**The deliverable is a FlyDSL kernel.** A PyTorch reimplementation of this
problem -- with or without `torch.compile` -- is the *baseline*, not a result.
It does not close this issue no matter how closely it matches the reference.

### Implementation contract (hard requirement)

The submitted kernel MUST be written in FlyDSL and go through the FlyDSL
compiler. Concretely, the kernel module has to actually use the DSL, e.g.:

```python
import flydsl.compiler as flyc
from flydsl.expr import arith as ea, gpu, vector as ev
```

and dispatch through `flyc.jit` / `flyc.kernel`. A file placed under
`FlyDSL/kernels/sol_execbench/` that only calls `torch.*` ops does NOT satisfy
this issue, even if it is numerically perfect and even if it is faster.

If you conclude the port is not expressible in FlyDSL today, say so explicitly
and report exactly which DSL capability is missing. That is a legitimate,
useful outcome. Silently substituting a PyTorch implementation is not.

### Baseline definition

The baseline is the sol-execbench PyTorch reference (`reference.py`) running
eager, with `torch.compile` OFF. Measure it yourself in this run, on the same
GPU and the same workload shapes you benchmark your kernel on. Do not carry a
number over from another run, another GPU, or the issue text.

### Performance target

Your FlyDSL kernel must be **at least 10% faster than that measured baseline**
-- `kernel_ms <= 0.90 x baseline_ms` -- on **at least 80% of the tested
workload shapes**.

### Correctness target

`max_abs_diff < 1e-2` versus the PyTorch reference on **all** tested shapes. A
faster kernel that fails this is a failure, not a tradeoff.

### What you must report

Report a per-shape table containing, for every tested shape: the measured
eager baseline ms, your FlyDSL kernel ms, the ratio, and win/loss. Then state
the win rate against the 80% bar. A result without the baseline column next to
the kernel column is not checkable and will not be credited.
